src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,486 @@
+    response_listener: Arc<ResponseListener>,
+}
+
+impl InMemoryRpcClient {
+    /// Creates a new RPC client for a given transport.
+    ///
+    /// # Arguments
+    ///
+    /// * `transport` - The uProtocol Transport Layer implementation to use for invoking service operations.
+    /// * `uri_provider` - The helper for creating URIs that represent local resources.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the generic RPC Response listener could not be
+    /// registered with the given transport.
+    pub async fn new(
+        transport: Arc<dyn UTransport>,
+        uri_provider: Arc<dyn LocalUriProvider>,
+    ) -> Result<Self, UStatus> {
+        let response_listener = Arc::new(ResponseListener {
+            pending_requests: Mutex::new(HashMap::new()),
+        });
+        transport
+            .register_listener(
+                &UUri::any(),
+                Some(&uri_provider.get_source_uri()),
+                response_listener.clone(),
+            )
+            .await?;
+
+        Ok(InMemoryRpcClient {
+            transport,
+            uri_provider,
+            response_listener,
+        })
+    }
+
+    #[cfg(test)]
+    fn contains_pending_request(&self, reqid: &UUID) -> bool {
+        self.response_listener.contains(reqid)
+    }
+}
+
+#[async_trait]
+impl RpcClient for InMemoryRpcClient {
+    async fn invoke_method(
+        &self,
+        method: UUri,
+        call_options: CallOptions,
+        payload: Option<UPayload>,
+    ) -> Result<Option<UPayload>, ServiceInvocationError> {
+        let message_id = call_options.message_id().unwrap_or_else(UUID::build);
+
+        let mut builder = UMessageBuilder::request(
+            method.clone(),
+            self.uri_provider.get_source_uri(),
+            call_options.ttl(),
+        );
+        builder.with_message_id(message_id.clone());
+        if let Some(token) = call_options.token() {
+            builder.with_token(token.to_owned());
+        }
+        if let Some(priority) = call_options.priority() {
+            builder.with_priority(priority);
+        }
+        if let Some(traceparent) = call_options.traceparent() {
+            builder.with_traceparent(traceparent);
+        }
+        let rpc_request_message = if let Some(pl) = payload {
+            let format = pl.payload_format();
+            builder.build_with_payload(pl.payload(), format)
+        } else {
+            builder.build()
+        }
+        .map_err(|e| ServiceInvocationError::InvalidArgument(e.to_string()))?;
+
+        let receiver = self
+            .response_listener
+            .try_add_pending_request(message_id.clone())?;
+        self.transport
+            .send(rpc_request_message)
+            .await
+            .map_err(|e| {
+                self.response_listener.remove_pending_request(&message_id);
+                e
+            })?;
+        match tokio::time::timeout(Duration::from_millis(call_options.ttl() as u64), receiver).await
```

Comment:
Given we don't use the error path's content consider:

```rust
if let Ok(Ok(response_message)) = tokio::time::timeout(Duration::from_millis(call_options.ttl() as u64), receiver).await {
  handle_response_message(response_message)
} else {
  self.response_listener.remove_pending_request(&message_id);
  Err(ServiceInvocationError::DeadlineExceeded)
}
```

Makes the line a little longer, so may then be useful to break up the latter half across a few lines, but presented for your consideration :slightly_smiling_face: